### DISCUSSION TITLE
Clipboard history

### DISCUSSION THREAD

**Discussion by Alarid** [johnlindquist/kit#290] [Q&A]
There is an entry in the "Help" section that I'm really interested in: "Clipboard History". I would like to bind a shortcut to access this feature directly, so I thought "Let's write a script for that !". But I have no idea how to do this 😅

I know there are globals like `copy` and `paste`, but I can't seem to find a way to access the clipboard history. Any tips on this?

**Comment by johnlindquist** [comment]
Clipboard history is included as a script in newer versions of .kenv. You can grab it here:


https://github.com/johnlindquist/kenv/blob/alpha/scripts/clipboard-history.js

**Reply by Alarid** [reply]
Amazing, thank you for your quick reply, that's perfect 👌

**Reply by hakubo** [reply]
looks like pasting image is not working? 
it's pasted as a path to .kenv tmp folder.

**Reply by johnlindquist** [reply]
@hakubo ah, ok, thanks for letting me know.

Fyi, the next build (maybe tomorrow?) will require an update to the clipboard-history script. 

* The `db` helper is changing to become a little simpler and more useful.
* The build will support multiple kit environments (kenvs), so I'm moving clipboard history to `~/.kit/tmp`

**Reply by horacioh** [reply]
seems that the link is outdated :(
does anybody knows the updated one ?

**Reply by Alarid** [reply]
You can use mine: [Install clipboard history](https://gist.githubusercontent.com/Alarid/cddb46003603566bec79434945786298/raw/b9e5925b78a940b375c396a6ae67f10507d1bc0c/clipboard-history.js)

**Reply by horacioh** [reply]
thanks @Alarid !! now I can officially uninstall Alfred from my computer :)
and of course thanks to @johnlindquist

**Reply by johnlindquist** [reply]
@Alarid @horacioh 

In the latest build, the clipboard history was refactored to be in-memory in Kit.app. Meaning it no longer writes to that history.json file. So it looks like this now:

```js
let value = await arg("Hit enter to paste", await getClipboardHistory())

await setSelectedText(value)
```

Or with previews:

```js
let history = await getClipboardHistory()
let historyWithPreviews = history.map(item => {
  item.preview = () => `<div class="text-xs p-5 font-mono">${item.value}</div>`
  return item
})
let value = await arg("Hit enter to paste", historyWithPreviews)

await setSelectedText(value)
```

I'm working on some APIs for "clear item", "clear history", etc and making an "official" script I'll include in `kit-examples`.
